Update NGINX SPDY patch for 1.9.15

felixbuenemann opened the pull request:
This updates the patch to work with NGINX 1.9.15 and also fixes a problem with the previous patch, that did not re-enable the `spdy_*` directives that were marked as deprecated by the http2 module as well as fixing a compile problem when the http2 module is enabled, but spdy is disabled.

felixbuenemann commented:
Btw. the patch also applies to 1.11.0 with some fuzz, but I have only tested it on 1.9.15.

felixbuenemann commented:
@jgrahamc If you want to keep the 1.9.7 patch around, I could also rename the updated patch and send a separate bugfix for the spdy directives in the old patch.

jgrahamc commented:
This is cool. Can you submit an updated patch instead so we can keep the old one around as well?

felixbuenemann committed:
Updated NGINX 1.9.15 SPDY Patch

This updates the 1.9.7 patch to work with NGINX 1.9.15 and also fixes
two problem with the previous patch, that did not re-enable spdy directives
that were marked as deprecated by the http2 module and failed to compile
when the http2 module was enabled and spdy was disabled.

felixbuenemann committed:
Fix broken spdy directives in nginx 1.9.7 patch

The various spdy configuration directives were still marked as
deprecated by the http v2 module in the existing patch.
Rhis change removes the deprecation so that the directives defined by
the spdy module can be used again.

felixbuenemann committed:
Add version number to NGINX 1.9.7 SPDY patch

This should make it clearer which version this patch is targeting, now
that there is another patch for 1.9.15.

felixbuenemann commented:
@jgrahamc I have separated the patch and fixed the non-working `spdy_*` directives in the 1.9.7 patch. I have also renamed the old patch to add the version number in order to avoid confusion.

felixbuenemann commented:
Btw. I tested the 1.9.15 patch against 1.11.0 and it works fine.

kn007 commented:
When I test this patch in 1.9.15 and 1.11.0, it's return same error:

```
src/http/modules/ngx_http_ssl_module.c: In function  ngx_http_ssl_npn_advertised :
src/http/modules/ngx_http_ssl_module.c:445:5: error: expected expression before  }  token
make[1]: *** [objs/src/http/modules/ngx_http_ssl_module.o] Error 1
```

frostieDE commented:
Works fine against nginx 1.11.0 and OpenSSL 1.0.2h (with ChaCha20 patch). I am building nginx as an Ubuntu 16.04 package.

Thanks for your work @felixbuenemann

felixbuenemann commented:
@kn007 I believe the extra block was added for C89 compatibility, where all variables have do be declared at the start of a block. I'm not sure why that leads to a compile error for you. One thing that looks problematic about the above code is that if only HTTP2 is compiled, there's an unmatched `else` branch. I think the block can be avoided by re-ordering the init and debug log code.

felixbuenemann commented:
@kn007 Oh, I just noticed I misread your first comment, you were talking about the case where only `--with-http_v2_module` is enabled. I'll look into that.

felixbuenemann committed:
Fix http2 without spdy compilation in 1.9.7 patch

Tis fixes a compile error caused by the 1.9.7 NGINX SPDY patch due
to an unmatched else branch, if the http_v2 module was enabled, but
the spdy module was disabled.

felixbuenemann commented:
@kn007 I have updated the 1.9.7 and 1.9.15 patches to fix the compile problem `--with-http_v2_module` but without `--with-http_spdy_module`.

Please do not use your fix removing lines 409 and 445, it is wrong and will only work by accident.

Thanks for the report!

felixbuenemann commented:
@jgrahamc I have kept the 1.9.7 patch fixes separate. I can combine them into a single commit and/or move them over to a separate PR if you prefer.

kn007 commented:
@felixbuenemann Got it, thank you. That's very kind of you.

felixbuenemann commented:
@jgrahamc It would be nice if the nginx spdy patch was mentioned in the `README.md`. I can't speak on behalf of CloudFlare, so it's best if you did that. A short paragraph linking to the blog post would probably be sufficient.

xetorixik commented:
The least I can do is write a big thank you for your time, effort and eventually the SPDY patch for Nginx 1.10 @felixbuenemann

kn007 commented:
nginx 1.11.1 test pass

Angristan commented:
Not working anymore for me : 
```
src/http/modules/ngx_http_ssl_module.c: In function ‘ngx_http_ssl_npn_advertised’:
src/http/modules/ngx_http_ssl_module.c:445:5: error: expected expression before ‘}’ token
     }
     ^
objs/Makefile:1170: recipe for target 'objs/src/http/modules/ngx_http_ssl_module.o' failed
make[1]: *** [objs/src/http/modules/ngx_http_ssl_module.o] Error 1
make[1]: *** Attente des tâches non terminées....
make[1]: Leaving directory '/opt/nginx-1.11.1'
Makefile:8: recipe for target 'build' failed
```

But still working with https://raw.githubusercontent.com/felixbuenemann/sslconfig/b8ebac6a337e8e4e373dfee76e7dfac3cc6c56e6/patches/nginx_1_9_15_http2_spdy.patch

felixbuenemann commented:
@Angristan Can't reproduce, works fine for me on 1.11.1 with only ssl module, ssl + http2, ssl + spdy and ssl + http2 + spdy. So you'd have to be a bit more specific.

Maybe the patch got applied incorrectly?

This script should work:

```sh
#!/bin/sh
set -e
git clone https://github.com/felixbuenemann/sslconfig.git
cd sslconfig
git checkout updated-nginx-1.9.15-spdy-patch
cd ..
git clone https://github.com/nginx/nginx.git
cd nginx
git checkout release-1.11.1
git apply ../sslconfig/patches/nginx_1_9_15_http2_spdy.patch
./auto/configure --prefix=/opt/nginx-1.11.1 \
  --with-http_ssl_module \
  --with-http_spdy_module \
  --with-http_v2_module
make
```

Angristan commented:
I'm using the patch this way : https://github.com/Angristan/nginx-autoinstall/blob/master/nginx-autoinstall.sh#L420

kn007 commented:
@Angristan Please using the laster version patch:
https://raw.githubusercontent.com/felixbuenemann/sslconfig/updated-nginx-1.9.15-spdy-patch/patches/nginx_1_9_15_http2_spdy.patch

Angristan commented:
@kn007 this one is working thx

raeesiqbal commented:
How do you test if it is actually using SPDY? Chrome doesn't support it anymore and Firefox says it's using `HTTP/1.1`

HansVanEijsden commented:
@raeesiqbal NGINX Amplify gives insights: https://amplify.nginx.com/

felixbuenemann commented:
You can test with Firefox by disabling http2 for the port in nginx or add another port with only spdy.

Angristan commented:
@raeesiqbal with https://www.ssllabs.com/ssltest/index.html and the openssl command from the blog article https://blog.cloudflare.com/open-sourcing-our-nginx-http-2-spdy-code/

felixbuenemann commented:
Firefox also add s a fake(?) header X-Firefox-Spdy in the developer console which shows the SPDY or HTTP2 version used.

raeesiqbal commented:
Oh yeah. It has a header `X-Firefox-SPDY: 3.1`, which means that it is working. Thanks everyone.

kn007 commented:
@raeesiqbal Testing like this: https://spdycheck.org/#kn007.net
Also can get protocols from ssllabs.com

Angristan commented:
@kn007 I think this site doesn't work very well, it says my website does not use HTTPS :') https://spdycheck.org/#angristan.fr

FernandoMiguel commented:
@Angristan cause it's HTTP2 and not 1.1 SDPY
![screenshot 2016-06-14 17 02 02](https://cloud.githubusercontent.com/assets/168366/16050023/d11b4a58-3251-11e6-98eb-536ccde3dd07.png)

HansVanEijsden commented:
But https://spdycheck.org/#www.weblogzwolle.nl uses SPDY and HTTP/2... also that same error. Same with all my other sites and servers.

kn007 commented:
@Angristan I don't know why, maybe because your site support `Forward Secrecy`?

Or something notice from spdycheck
```
Notice:
Many of the SPDY implementations out there are starting to switch from NPN to APLN negotiation. SPDYCheck does not yet support APLN negotiation. Sites like Facebook and Twitter have different SPDY implementations deployed behind load balancers, which can lead to some results saying SPDY is supported while others saying it is not. If you receive a results that you know is incorrect. Try running the check again.
```

I test your site in ssllab, it was return `h2 spdy/3.1 http/1.1`.

So, maybe spdycheck doesn't work very well.

Angristan commented:
@FernandoMiguel 
```
root@lyra ~# openssl s_client -connect angristan.fr:443 -nextprotoneg ''
CONNECTED(00000003)
Protocols advertised by server: h2, spdy/3.1, http/1.1
```
@kn007 Yep that's maybe ALPN

kn007 commented:
@HansVanEijsden Maybe because you and @Angristan website support Forward Secrecy.

HansVanEijsden commented:
@kn007 that must be it. Thanks.

kn007 commented:
@Angristan my website also support ALPN and NPN, I guess maybe because `Forward Secrecy`.

kn007 commented:
@Angristan @HansVanEijsden Got the reseon, spdycheck not support ECC CA. I test few site ECC CA website, return failed.

felixbuenemann commented:
You could use NGINX 1.11+ multiple certificate support to supply both ECC and RSA certificates, so that older clients can connect to your server.

Angristan commented:
@felixbuenemann Thoses who do not support ECC certs are thoses who do not support TLS 1.2 so I prefer to avoid them ^^

wangqiliang commented:
https://wangqiliang.com/ben-zhan-qian-duan-yi-dai-li-fu-wu-qi-chu-cuo-zan-che-xia/

kn007 commented:
@wangqiliang 请用英文。而且不要以链接方式来评论。你可以尝试使用Google Translate来附上你想评论的内容。

wangqiliang sending a feedback about when using IE9 to visit website with nginx 1.11.1&spdy patch, sometime will return ~4-bits anomaly contents and TTFB has reached 30000ms.

When not using `spdy patch`, it was quite normal.

wangqiliang commented:
@kn007 Sorry, it is my fault.
I am sending a feedback that when I access the port 80 which I set a `spdy` parameter in the ngx/1.11.1 with patch, there is unexpected ~4-bit contents output with a timeout (in Firefox 46, 30000ms) .
( I should correct a mistake. It was not a 30000-ms-TTFB but a ERR_TIME_OUT.)
when I am not using this patch, it returns 301 to HTTPS normally.

kn007 commented:
@wangqiliang It's OK. Please excuse any mistakes for my english.

felixbuenemann commented:
@wangqiliang You can only use spdy on port 80, if you have a load balancer like haproxy in front of your nginx which handles SSL termination and ALPN/NPN negotiation. If you use spdy or http2 on a port without ssl in NGINX, it will always use that protocol without negotiation.

So the simple answer is, don't ser spdy on port 80, only use it on 443.

wangqiliang commented:
@felixbuenemann
It's so kind of you.
I will check for it later.

wangqiliang commented:
@felixbuenemann @kn007 From Wikepedia,
> For use within HTTPS, SPDY needs the TLS extension Next Protocol Negotiation (NPN),[24] thus browser and server support depends on the HTTPS library.

So that is my fault. Thanks for your kindness!

kn007 commented:
@felixbuenemann nginx 1.11.3 test failud
```
src/stream/ngx_stream_proxy_module.c:542:6: error:  ngx_stream_upstream_t  has no member named  ssl_name’
     u->ssl_name = uscf->host;
      ^
```

kn007 commented:
Do not work in 1.11.5

fmunteanu commented:
@kn007, that repo patch does not work with you add dynamic modules.

natnet00 commented:
@felixbuenemann
Your patch works flawlessly up to version 1.11.4
But starting with nginx 1.11.5 the patch unfortunately gives errors... would be great if this could be fixed.

felixbuenemann commented:
@natnet00 Have you tried the patch linked by kn007 above?

natnet00 commented:
@felixbuenemann nope have not tried it yet

lenovouser commented:
@felixbuenemann both also doesn't work with 1.11.6 :/

Injust commented:
@lenovouser The patch linked by kn007 indeed works for 1.11.6.

lenovouser commented:
@Injust nope, see cujanovic/nginx-http2-spdy-patch/issues/1 and 

![](https://i.lenovouser.me/yoAeD3iElt.png)

kn007 commented:
@lenovouser I am upgrade Nginx to 1.11.6 withi this path last week, it's working.

kn007 commented:
![20161124072106](https://cloud.githubusercontent.com/assets/6196903/20581857/a19ec936-b216-11e6-888c-36389b71c463.jpg)
@lenovouser

lenovouser commented:
@kn007 weird, what NGINX source are you pulling from? I use `https://nginx.org/download/nginx-1.11.6.tar.gz`

kn007 commented:
@lenovouser Yes, i am using `http://nginx.org/download/nginx-1.11.6.tar.gz`

lenovouser commented:
Which one of the two files are you using? Because both are failing for me

kn007 commented:
@lenovouser 
wget -c http://nginx.org/download/nginx-1.11.6.tar.gz
tar zxvf nginx-1.11.6.tar.gz && cd nginx-1.11.6/
wget -c https://raw.githubusercontent.com/cujanovic/nginx-http2-spdy-patch/master/nginx-spdy-1.11.5%2B.patch
patch -p1 < nginx-spdy-1.11.5+.patch

kn007 commented:
![2016-11-24_075357](https://cloud.githubusercontent.com/assets/6196903/20582330/4230439e-b21b-11e6-95fc-f716f9224a00.png)

lenovouser commented:
So weird. Works on my local machine, but the Docker container I am trying to build it in fails. Well. Seems like I have to go find out the bug myself. Thanks for your help!

kn007 commented:
@lenovouser OK, good luck!

lenovouser commented:
So, I found the issue. In case anyone reads this in the future and you're using Alpine Linux, just do a `apk add patch`. That will update the `patch` binary. For some reason the one installed by default is buggy 😞

felixbuenemann commented:
@lenovouser It's probably using a busybox version of patch by default.

pankajslingwal commented:
> So, I found the issue. In case anyone reads this in the future and you're using Alpine Linux, just do a `apk add patch`. That will update the `patch` binary. For some reason the one installed by default is buggy 😞

dude, you saved my day :)
